Refactor selection in SelectingItemsControl and TreeView

grokys opened the pull request:
## What does the pull request do?

*NOTE: This PR targets `master` and includes all the changes in #3469 #3470 #3498 #3509 and #3445. **MERGE THIS PR ON APPROVAL, NOT THOSE***

This PR is the final part of #3469. It refactors `SelectingItemsControl` and `TreeView` to use `SelectionModel` (based on #3469 #3470 #3498 #3509).

In #3469 I mention that we may end up removing `SelectedItems`, I did't end up doing this:

1. If the user doesn't want their VM layer to have a reference to `Avalonia.Controls` then they might prefer to use `SelectedItems`
2. The logic to sync them wasn't too complex

I've ended up not even marking `SelectedItems` as `[Obsolete]` because of point 1.

The `SelectionModel` is exposed via the `Selection` property which can also be bound. If the user accesses the `SelectedItems` property then the `SelectionModel` and `SelectedIndexes` are kept in sync.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [x] Submitted a PR to avaloniaui.net: https://github.com/AvaloniaUI/avaloniaui.net/pull/111

## Breaking changes

Changes selection behavior in a couple of cases due to differences in behavior between our previous implementation and `SelectionModel`. The changes shouldn't be too serious as the choices we made previously were contrary to WPF etc:

- Moving the selected item now clears the selection
- Replacing the selected item now clears the selection
- Removing the selected item with `AlwaysSelect` now selects the first item, not the next item

## Fixed issues

Fixes #2569 
Fixes #3497 
Fixes #3448 
Fixes #3450 
Fixes #3576

MarchingCube committed:
Failing unit tests for SelectingItemsControl events

MarchingCube committed:
Verify that removing items raises events as well.

grokys committed:
Swap order of parameters to event args.

As described in #3437 the argument ordering was different to WPF/UWP causing bugs in some code ported from WPF/UWP. Use the same argument ordering as WPF/UWP.

Fixes #3437

MarchingCube committed:
Fix unit tests.

MarchingCube committed:
Merge branch 'master' into selecting-items-control-events

grokys committed:
Ported SelectionModel and friends from WinUI.

grokys committed:
Refactor IndexPath.

- #nullable enable
- Make it a `readonly struct`
- Most of the time it will only hold a single `int`, so optimize for the common case by having an `int _index` field
- Make `_path` an array rather than a list as it will be fixed-size
- Implement equality
- Implement operators

grokys committed:
Add nullability annotations to SelectionModel etc.

And ran a few of VS' lightbulb suggestions.

grokys committed:
Add `IndexRange` list add/remove methods.

Add or remove index ranges from a list of index ranges, merging and splitting ranges as required.

grokys committed:
Add `IndexRange` list add/remove methods.

Add or remove index ranges from a list of index ranges, merging and splitting ranges as required.

grokys committed:
Added SelectionModel changed args.

`SelectionModel` as ported from WinUI has no information about what changed in a `SelectionChanged` event. This adds that information along with unit tests.

grokys committed:
Added SelectionModel changed args.

`SelectionModel` as ported from WinUI has no information about what changed in a `SelectionChanged` event. This adds that information along with unit tests.

grokys committed:
Added some failing tests.

That demonstrate some problems with the `SelectionModel` change notifications found so far.

grokys committed:
Added some failing tests.

That demonstrate some problems with the `SelectionModel` change notifications found so far.

grokys committed:
Refactor of SelectionModel change notifications.

To address issues found.

grokys committed:
Refactor of SelectionModel change notifications.

To address issues found.

grokys committed:
Make comparing IndexPath with null do something useful.

grokys committed:
Added SelectionMode.AutoSelect.

grokys committed:
Lazily create the selected/deselected lists.

grokys committed:
Lazily create the selected/deselected lists.

grokys committed:
Use SelectedItems for change event args.

grokys committed:
Use SelectedItems for change event args.

grokys committed:
Added SelectionModel.RetainSelectionOnReset.

grokys committed:
Ported fix and test from WinUI.

https://github.com/microsoft/microsoft-ui-xaml/pull/1922

grokys committed:
Add nullable annotations.

grokys committed:
Handle null SelectionModel.Source.

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Don't reset selection if source hasn't changed.

Also remove some `?.` operators that aren't needed.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Fixed out of bounds in SelectionNode.

grokys committed:
Fix bad merge.

grokys committed:
Handle RetainSelectionOnReset w/ null source.

grokys committed:
Expose API for batch updates.

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Merge branch 'feature/selectionmodel-reset-handling' into feature/selectionmodel-autoselect

grokys committed:
Added ISelectionModel.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Merge branch 'feature/selectionmodel-reset-handling' into feature/selectionmodel-autoselect

grokys committed:
Added AutoSelect to ISelectionModel.

grokys committed:
Added ISelectionModel.Update().

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Merge branch 'feature/selectionmodel-reset-handling' into feature/selectionmodel-autoselect

grokys committed:
Added failing test.

Clearing a nested selection doesn't raise `SelectionChanged`.

grokys committed:
Fix clearing nested selection not raising SelectionChanged.

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Merge branch 'feature/selectionmodel-reset-handling' into feature/selectionmodel-autoselect

grokys committed:
Clean up when node removed.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Fix SelectionNode.Cleanup.

- Removed disposed child nodes
- Don't dispose child node with descendent selection

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Merge branch 'feature/selectionmodel-reset-handling' into feature/selectionmodel-autoselect

grokys committed:
ISelectionModel implements INotifyPropertyChanged.

This will be needed for monitoring the `AnchorIndex` in order to auto-scroll.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Remove UWP stubs in unit tests.

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

# Conflicts:
#	tests/Avalonia.Controls.UnitTests/SelectionModelTests.cs

grokys committed:
Merge branch 'feature/selectionmodel-reset-handling' into feature/selectionmodel-autoselect

grokys committed:
Fixed SelectionModelChildrenRequestedEventArgs returning incorrect so...

…urce index

fixed sourceindex childrenrequested args

Port of https://github.com/microsoft/microsoft-ui-xaml/commit/ee03bace144e3a37650df0536955f5f658dd2aad

grokys committed:
Merge branch 'feature/selectionmodel' into feature/selectionmodel-changed-notifications

grokys committed:
Merge branch 'feature/selectionmodel-changed-notifications' into feature/selectionmodel-reset-handling

grokys committed:
Merge branch 'feature/selectionmodel-reset-handling' into feature/selectionmodel-autoselect

grokys committed:
Don't only select leaf nodes from SelectRange.

Related issues:

https://github.com/microsoft/microsoft-ui-xaml/issues/1969
https://github.com/microsoft/microsoft-ui-xaml/issues/1984

grokys committed:
Changed IsSelected semantics.

Move old `IsSelected` semantics to `IsSelectedWithPartial` and add a new `IsSelected` method which checks for direct selection.

grokys committed:
Merge branch 'pr/3435' into refactor/use-selectionmodel

grokys committed:
Added SelectedItemsSync.

To sync between an `ISelectionModel` and a `SelectedItems` collection.

grokys committed:
Use SelectionModel in SelectingItemsControl.

grokys committed:
Use SelectionModel in TreeView.

grokys committed:
Make ScrollIntoView accept an index, not an item.

Fixes #2569.

grokys committed:
Merge branch 'master' into refactor/use-selectionmodel

grokys commented:
Actually just discovered a problem here, back to WIP...

grokys committed:
TryGetValue doesn't allow null as the value.

MarchingCube commented on src/Avalonia.Controls/Primitives/SelectingItemsControl.cs:
```diff
@@ -190,32 +198,110 @@ public object SelectedItem
-                UnsubscribeFromSelectedItems();
-                _selectedItems = value ?? new AvaloniaList<object>();
-                SubscribeToSelectedItems();
+                        if (_selectedIndex != selectedIndex)
```
Could use `SetAndRaise`? Also it seems that you are first raising changed event and then changing the value.

grokys replied:
The thing is, `SelectedIndex` goes straight to the `SelectionModel`, this field is only used to store the old value for raising notifications. I guess I should probably call it `_oldSelectedIndex`...

MarchingCube replied:
Ah, I didn't get to that part yet :)

grokys commented:
Added some initial documentation on this feature to https://github.com/AvaloniaUI/avaloniaui.net/pull/111

MarchingCube commented on src/Avalonia.Controls/SelectionModel.cs:
```diff
@@ -0,0 +1,860 @@
+                    }
+
+                    // Instead of creating a dumb vector that takes up the space for all the selected items,
+                    // we create a custom VectorView implimentation that calls back using a delegate to find 
```
implementation.

grokys replied:
https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/SelectionModel.cpp#L174 ;)

Also `VectorView` is what C++ calls `IReadOnlyList<>` - should probably change that too.

MarchingCube commented on src/Avalonia.Controls/SelectedItems.cs:
```diff
@@ -0,0 +1,49 @@
+        {
+            for (var i = 0; i < Count; ++i)
+            {
+                yield return this[i];
```
I am already getting flashbacks to visual tree operations using `yield` and taking 10x longer than they should :) Can you just implement a simple struct enumerator?

grokys replied:
I did consider this, but `SelectedItems` is internal and only exposed to the outside world as an `IReadOnlyList<T>`. We could expose `SelectedItems` but I was trying to stick to WinUI's public API. I think probably the best thing to do here is use the indexer if you want to avoid allocations.

grokys replied:
Again, might make sense to address perf issues in another PR rather than obfuscate this one.

MarchingCube commented on src/Avalonia.Controls/SelectionModel.cs:
```diff
@@ -0,0 +1,860 @@
+
+        private void SelectRangeImpl(IndexPath start, IndexPath end, bool select)
+        {
+            var winrtStart = start;
```
winrt ?

grokys replied:
Yeah, comes from WinUI. I try to avoid renaming stuff when porting to make porting subsequent changes easier, however in this case it might be worth changing it because I agree, this really stands out. What do you think?

MarchingCube replied:
Well, from what I understood is that our `SelectionModel` differs from the original anyway so I figured we might want to unify some names. 

But I have no problem with keeping `winrt` although it looks a bit weird in our context.

grokys replied:
I think I'd rather keep the naming from upstream to ease porting of subsequent changes.

MarchingCube commented on src/Avalonia.Controls/SelectionModelChangeSet.cs:
```diff
@@ -0,0 +1,170 @@
+{
+    internal class SelectionModelChangeSet
+    {
+        private List<SelectionNodeOperation> _changes;
```
`readonly`.

MarchingCube commented on src/Avalonia.Controls/SelectionModelChangeSet.cs:
```diff
@@ -0,0 +1,170 @@
+        {
+            static int GetCount(SelectionNodeOperation info) => info.DeselectedCount;
+            static List<IndexRange>? GetRanges(SelectionNodeOperation info) => info.DeselectedRanges;
+            return GetIndexAt(infos, index, GetCount, GetRanges);
```
Since C# is such a great language you have to write something like ` GetIndexAt(infos, index, i => GetCount(i), i => GetRanges(i))` to avoid allocating new delegates on every call :)

[Sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs7o64AcpsANwCGGGAxR0AKjADOGABSEArAB5sAOwwA+OgHMYGMVACUPTu0pc6AX3PcrXWhLoBZBSYfXL164VJIdJoYdADihkwQAK5aHnQAvLrKANz21mlcMvIKcYlhEdGxJiapjpx2ZV5O9ISSrsRx9j6+nP6BwfkYkTGKnnkpGZyD0nKK4V2FGCX2FTZAA===)

grokys replied:
Ah yeah, this always trips me up - it just feels like the non-lambda syntax should be better!

kronic replied:
As far as I know, no allocation occurs for local static functions

grokys committed:
Fix comment.

Fix spelling, and `VectorView` is `IReadOnlyList` in C#.

grokys committed:
Tweak SelectionModelChangeSet.

- Make field readonly
- Use lambda syntax so C# caches delegates

grokys committed:
Merge branch 'master' into refactor/use-selectionmodel

grokys committed:
Merge branch 'master' into refactor/use-selectionmodel

MarchingCube committed:
Merge branch 'master' into refactor/use-selectionmodel

MarchingCube committed:
Merge branch 'master' into refactor/use-selectionmodel

grokys committed:
Merge branch 'master' into refactor/use-selectionmodel

grokys committed:
Added some docs for ISelectionModel.

grokys commented on src/Avalonia.Controls/ISelectionModel.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace Avalonia.Controls
+{
+    public interface ISelectionModel : INotifyPropertyChanged
```
Added some basic docs. Once the WinUI docs for this class are published, will update it with those.

grokys commented on src/Avalonia.Controls/Utils/SelectedItemsSync.cs:
```diff
@@ -0,0 +1,227 @@
+
+            try
+            {
+                var deselected = e.DeselectedItems.ToList();
```
Would it make sense to address perf issues like this in another PR? This one is already hard enough to review...

MarchingCube committed:
Merge branch 'master' into refactor/use-selectionmodel

grokys committed:
Merge branch 'master' into refactor/use-selectionmodel

grokys committed:
Allow SelectionNode children to change.

Make `SelectionModelChildrenRequestedEventArgs.Children` an observable, so that the we can react to the children collection object changing, as well as the children inside the collection changing.

Upstream issue: https://github.com/microsoft/microsoft-ui-xaml/issues/2404

grokys committed:
Merge branch 'master' into refactor/use-selectionmodel

igorstefus commented:
After fixing problem with `SelectionNode` children we should be good to go. I did some testing with custom control with tree structure like source and selection mechanics looks to be fine. We will do some more testing but it looks like it is ready for merging.

igorstefus reviewed (approved):
After fixing problem with SelectionNode children we should be good to go. I did some testing with custom control with tree structure like source and selection mechanics looks to be fine. We will do some more testing but it looks like it is ready for merging.

grokys committed:
Use correct unit test AppBuilder.

A reference was added to `Avalonia.ReactiveUI` from `Avalonia.Controls.UnitTests` which caused the incorrect `AppBuilder` to be resolved.
